File: services/snowflake/service/snowflake.go
package service

import (
	"fmt"
	"sync"
	"time"
)

// https://en.wikipedia.org/wiki/Snowflake_ID
const (
	epoch          int64 = 1735689601 // Jan 1st, 2025 00:00:01 GMT+0000
	machineIdBits  uint  = 10         // machine id bits
	sequenceBits   uint  = 12         // sequence bits
	machineIdShift uint  = sequenceBits
	timestampShift uint  = sequenceBits + machineIdBits
	maxMachineId   int64 = -1 ^ (-1 << machineIdBits)
	maxSequence    int64 = -1 ^ (-1 << sequenceBits)
)

type Snowflake struct {
	mu        sync.Mutex
	machineId int64
	lastTime  int64
	sequence  int64
}

func NewSnowflake(machineId int64) (*Snowflake, error) {
	if machineId < 0 || machineId > maxMachineId {
		return nil, fmt.Errorf("machine ID must be between 0 and %d", maxMachineId)
	}

	return &Snowflake{
		machineId: machineId,
		lastTime:  -1,
		sequence:  0,
	}, nil
}

func (s *Snowflake) Snowflake() int64 {
	s.mu.Lock()
	defer s.mu.Unlock()

	now := time.Now().UnixMilli()
	if now == s.lastTime {
		// Increment sequence if in the same millisecond
		s.sequence = (s.sequence + 1) & maxSequence
		if s.sequence == 0 {
			// Sequence overflow, wait for the next millisecond
			for now <= s.lastTime {
				now = time.Now().UnixMilli()
			}
		}
	} else {
		// Reset sequence for a new millisecond
		s.sequence = 0
	}
	s.lastTime = now

	// Generate the ID
	id := ((now - epoch) << timestampShift) |
		(s.machineId << machineIdShift) |
		s.sequence

	return id
}

File: lib/snowflake.ts
const EPOCH = 1735689601; // Jan 1st, 2025 00:00:01 GMT+0000
const MACHINE_ID_BITS = 10; // machine id bits
const SEQUENCE_BITS = 12; // sequence bits
const MACHINE_ID_SHIFT = SEQUENCE_BITS;
const TIMESTAMP_SHIFT = SEQUENCE_BITS + MACHINE_ID_BITS;
const MAX_MACHINE_ID = -1 ^ (-1 << MACHINE_ID_BITS);
const MAX_SEQUENCE = -1 ^ (-1 << SEQUENCE_BITS);

class Snowflake {
  private static instance: Snowflake;

  static getInstance() {
    if (!this.instance) {
      this.instance = new Snowflake();
    }

    return this.instance;
  }

  private machineId = 0;
  private lastTime = 0;
  private sequence = 0;

  generate(): number {
    let now = new Date().getTime();

    if (now === this.lastTime) {
      this.sequence = (this.sequence + 1) & MAX_SEQUENCE;

      // handle sequence overflow
      if (this.sequence === 0) {
        // block until the next millisecond
        while (now <= this.lastTime) {
          now = new Date().getTime();
        }
      }
    } else {
      this.sequence = 0;
    }

    this.lastTime = now;

    return Number(
      (BigInt(now - EPOCH) << BigInt(TIMESTAMP_SHIFT)) |
        (BigInt(this.machineId) << BigInt(MACHINE_ID_SHIFT)) |
        BigInt(this.sequence)
    );
  }
}

export const snowflake = Snowflake.getInstance();
